src/strawberry_sqlalchemy_mapper/relay.py

```diff
@@ -0,0 +1,427 @@
+    *,
+    session: Session,
+    info: Optional[Info] = ...,
+    required: Literal[True],
+    filter_perms: bool = False,
+) -> AwaitableOrValue[_T]:
+    ...
+
+
+def resolve_model_node(
+    source,
+    node_id,
+    *,
+    session: Session,
+    info: Optional[Info] = None,
+    required=False,
+    filter_perms=False,
+):
+    """
+    Resolve model node.
+
+    Args:
+    ----
+        source:
+            The source model or the model type that implements the `Node` interface
+        node_id:
+            The node it to retrieve the model from
+        session:
+            A sqlalchemy session object.
+        info:
+            Optional gql execution info. Make sure to always provide this or
+            otherwise, the queryset cannot be optimized in case DjangoOptimizerExtension
+            is enabled. This will also be used for `is_awaitable` check.
+        required:
+            If the return value is required to exist. If true, `qs.get()` will be
+            used, which might raise `model.DoesNotExist` error if the node doesn't
+            exist. Otherwise, `qs.first()` will be used, which might return None.
+
+    Returns:
+    -------
+        The resolved node, already prefetched from the database
+
+    """
+    from strawberry_sqlalchemy_mapper.mapper import StrawberrySQLAlchemyType
+
+    definition = StrawberrySQLAlchemyType[Any].from_type(source, strict=True)
+    model = definition.model
+
+    if isinstance(node_id, relay.GlobalID):
+        node_id = node_id.node_id
+
+    attrs = cast(relay.Node, source).resolve_id_attr().split("|")
+    converters = [getattr(model, attr).type.python_type for attr in attrs]
+
+    filter = dict(
+        zip(
+            cast(relay.Node, source).resolve_id_attr().split("|"),
+            (converters[i](nid) for i, nid in enumerate(node_id.split("|"))),
+        )
+    )
+
+    try:
+        return session.get(model, filter)
+    except NoResultFound:
+        if not required:
+            return None
+
+        raise
+
+
+def resolve_model_id_attr(source: Type) -> str:
+    """
+    Resolve the model id attribute.
+
+    In case of composed primary keys, those will be returned separated by a `|`.
+    """
+    try:
+        id_attr = super(source, source).resolve_id_attr()
+    except NodeIDAnnotationError:
+        from strawberry_sqlalchemy_mapper.mapper import StrawberrySQLAlchemyType
+
+        definition = StrawberrySQLAlchemyType[Any].from_type(source, strict=True)
+        id_attr = "|".join(
+            key.name for key in sqlalchemy_inspect(definition.model).primary_key
+        )
+
+    return id_attr
+
+
+def resolve_model_id(
+    source: type,
+    root: Any,
+    *,
+    info: Optional[Info] = None,
+) -> AwaitableOrValue[str]:
+    """
+    Resolve the model id.
+
+    In case of composed primary keys, those will be returned separated by a `|`.
+    """
+    id_attr = cast(relay.Node, source).resolve_id_attr().split("|")
```

Comment:
Let's find a way to cache this on the type later, no need to re-compute on every call